File: app/src/main/java/com/example/taskdone/Finanzas/PrincipalFragment.java
package com.example.taskdone.Finanzas;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;

import com.example.taskdone.Acceso.ActivityLogin;
import com.example.taskdone.DataBase;
import com.example.taskdone.Preferences;
import com.example.taskdone.UsuarioSingleton;
import com.example.taskdone.Utils;
import com.example.taskdone.DatePickerFragment;
import com.example.taskdone.R;
import com.example.taskdone.databinding.FragmentFinanzasPrincipalBinding;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Objects;


@RequiresApi(api = Build.VERSION_CODES.O)
public class PrincipalFragment extends Fragment {

    private FragmentFinanzasPrincipalBinding binding;

    NavController navController;

    DataBase dataBase;
    ArrayList<String> tags;

    ArrayList<String> monedas = new ArrayList<>();
    ArrayList<Float> cantidades = new ArrayList<>();
    ArrayList<String> simbolos = new ArrayList<>();

    @SuppressLint({"SetTextI18n", "UseCompatLoadingForDrawables", "RestrictedApi"})
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentFinanzasPrincipalBinding.inflate(inflater, container, false);
        navController = NavHostFragment.findNavController(this);

        dataBase = new DataBase(requireContext());

        //Obtengo las monedas del usuario con sus cantidades y simbolos

        Cursor data = dataBase.getMonedasByUserId(UsuarioSingleton.getInstance().getID());
        while (data.moveToNext()) {
            monedas.add(data.getString(1));
            cantidades.add(data.getFloat(2));
            simbolos.add(data.getString(3));
        }
        if(monedas.isEmpty()){
            navController.navigate(R.id.crearMonedaFragment);
        }
        else{
            llenarLayoutMonedas();
        }

        ArrayAdapter<String> spinnerArrayAdapter = new ArrayAdapter<String>(requireActivity(), android.R.layout.simple_spinner_dropdown_item, monedas){
            @RequiresApi(api = Build.VERSION_CODES.O)
            public View getView(int position, View convertView, @NonNull ViewGroup parent) {
                View v = super.getView(position, convertView, parent);

                ((TextView) v).setTextSize(20);
                ((TextView) v).setTextAlignment(View.TEXT_ALIGNMENT_CENTER);

                return v;
            }
        };

        spinnerArrayAdapter.setDropDownViewResource(R.layout.spinner_dropdown);
        binding.spinnerMoneda.setAdapter(spinnerArrayAdapter);
        binding.spinnerMoneda.setBackground(getResources().getDrawable(R.drawable.fondo_blanco_redondeado));
        binding.spinnerMoneda.setGravity(Gravity.CENTER);
        binding.spinnerMoneda.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                Preferences.savePreferenceString(requireContext(), Integer.toString(binding.spinnerMoneda.getSelectedItemPosition()), "gasto_moneda_index");
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        binding.agregarTag.setOnClickListener(v -> this.irATags());

        dataBase = new DataBase(requireContext());

        binding.ok.setOnClickListener(v -> {
            try {
                addData();
            } catch (ParseException e) {
                e.printStackTrace();
            }
        });

        tags = new ArrayList<>();

        if(getArguments() != null){
            if(getArguments().getString("scroll_tags") != null) {
                scrollToTag();
            }
        }

        binding.editFecha.setOnClickListener(v -> showDatePickerDialog());

        binding.agregarMoneda.setOnClickListener(v -> irACrearMoneda());

        binding.editCantidad.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                verificarCantidad();
                Preferences.savePreferenceString(requireContext(), binding.editCantidad.getText().toString(), "gasto_cantidad");
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });
        binding.editMotivo.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                Preferences.savePreferenceString(requireContext(), binding.editMotivo.getText().toString(), "gasto_motivo");
            }

            @Override
            public void afterTextChanged(Editable editable) {
            }
        });

        binding.checkIngreso.setOnCheckedChangeListener((compoundButton, b) -> guardarIngresoPendiente(b));

        cargarGastoPendiente();

        binding.logout.setOnClickListener(v -> popupCerrarSesion());

        binding.ayuda.setOnClickListener(v -> Utils.popupAyuda(requireContext(), requireActivity(), new ArrayList<>(Arrays.asList(getResources().getString(R.string.ayuda_principal_1), getResources().getString(R.string.ayuda_principal_2)))));

        return binding.getRoot();
    }

    void guardarIngresoPendiente(boolean b){
        Preferences.savePreferenceString(requireContext(), "0", "gasto_ingreso");
        if(b){
            Preferences.savePreferenceString(requireContext(), "1", "gasto_ingreso");
        }
    }

    private void irACrearMoneda(){
        Preferences.savePreferenceString(requireContext(), ""+R.id.principalFragment, "id_fragment_anterior");
        navController.navigate(R.id.crearMonedaFragment);
    }

    void irATags() {
        Preferences.savePreferenceString(requireContext(), ""+R.id.principalFragment, "id_fragment_anterior");
        navController.navigate(R.id.action_principalFragment_to_tagsFragment);
    }

    void cargarGastoPendiente(){
        String fecha = Preferences.getPreferenceString(requireContext(), "gasto_fecha");
        String moneda_index =Preferences.getPreferenceString(requireContext(), "gasto_moneda_index");
        String cantidad = Preferences.getPreferenceString(requireContext(), "gasto_cantidad");
        String motivo = Preferences.getPreferenceString(requireContext(), "gasto_motivo");
        String ingreso = Preferences.getPreferenceString(requireContext(), "gasto_ingreso");
        String tags_concatenados = Preferences.getPreferenceString(requireContext(), "gasto_tags");

        binding.editFecha.setText(Utils.calendarToString(Calendar.getInstance()));
        if(!fecha.equals("")) {
            binding.editFecha.setText(fecha);
        }
        if(!moneda_index.equals("")){
            binding.spinnerMoneda.setSelection(Integer.parseInt(moneda_index));
        }
        if(!cantidad.equals("")) {
            binding.editCantidad.setText(cantidad);
        }
        if(!motivo.equals("")) {
            binding.editMotivo.setText(motivo);
        }
        if(ingreso.equals("1")) {
            binding.checkIngreso.setChecked(true);
        }
        StringBuilder tags_seleccionados = new StringBuilder();
        if(!tags_concatenados.equals("")) {
            tags.addAll(Arrays.asList(tags_concatenados.split(", ")));
            tags_seleccionados.append(Utils.arrayListToString(tags));

        }
        else{
            tags_seleccionados.append(getResources().getString(R.string.sin_seleccionados));
        }
        binding.txtTagSeleccionados.setText(tags_seleccionados);
    }


    @SuppressLint("SetTextI18n")
    private void llenarLayoutMonedas(){
        while(binding.layoutMonedas.getChildCount()!=0) {
            binding.layoutMonedas.removeViewAt(0);
        }
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        for(int x=0;x!=monedas.size();x++){
            @SuppressLint("InflateParams") View view = inflater.inflate(R.layout.item_moneda_cantidad, null);
            @SuppressLint("CutPasteId") TextView moneda = view.findViewById(R.id.moneda);
            @SuppressLint("CutPasteId") TextView cantidad = view.findViewById(R.id.cantidad);
            moneda.setText(monedas.get(x));
            cantidad.setText(simbolos.get(x)+" "+Utils.formatoCantidad(cantidades.get(x)));
            binding.layoutMonedas.addView(view);
        }
    }

    private void scrollToTag(){
        binding.scrollview.post(() -> binding.scrollview.fullScroll(ScrollView.FOCUS_DOWN));
    }


    private void verificarCantidad(){
            if (!binding.editCantidad.getText().toString().equals("") && !binding.editCantidad.getText().toString().equals("0")) {
                if (binding.editCantidad.getText().toString().startsWith("0")) {
                    binding.editCantidad.setText(binding.editCantidad.getText().toString().substring(1));
                    binding.editCantidad.setSelection(binding.editCantidad.length());
                }
            }

    }

    private void showDatePickerDialog() {
        DatePickerFragment newFragment = DatePickerFragment.newInstance((datePicker, year, month, day) -> {

            String selectedDate = year + "/" + Utils.twoDigits(month + 1) + "/" + Utils.twoDigits(day);
            final Calendar h = Calendar.getInstance();
            h.set(year,month,day);
            h.add(Calendar.DAY_OF_YEAR,1);
            binding.editFecha.setText(selectedDate);
            Preferences.savePreferenceString(requireContext(), binding.editFecha.getText().toString(), "gasto_fecha");
            }

        );
        newFragment.show(Objects.requireNonNull(requireActivity()).getSupportFragmentManager(), "datePicker");
    }

    public void addData() throws ParseException {
        if (Utils.fechaMayorQueHoy(binding.editFecha.getText().toString())) {
            Toast.makeText(requireContext(), R.string.error_fecha_futura, Toast.LENGTH_SHORT).show();
        }
        else {
            if (binding.editCantidad.getText().toString().equals("") || binding.editCantidad.getText().toString().equals("0")) {
                Toast.makeText(requireContext(), R.string.error_cantidad_null, Toast.LENGTH_SHORT).show();
            } else {
                String ingreso = "0";
                if (binding.checkIngreso.isChecked()) {
                    ingreso = "1";
                }
                boolean insertData = dataBase.addGastos(binding.editFecha.getText().toString(), binding.spinnerMoneda.getSelectedItem().toString(), Float.parseFloat(binding.editCantidad.getText().toString()), binding.editMotivo.getText().toString(), tags, ingreso);

                if (insertData) {
                    Toast.makeText(requireContext(), R.string.guardado_exito, Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(requireContext(), R.string.error_guardado, Toast.LENGTH_SHORT).show();
                }
                cleanAndUpdate();
            }
        }
    }


    @SuppressLint("SetTextI18n")
    private void cleanAndUpdate(){
        Preferences.deletePreferencesGastoPendiente(requireContext());
        binding.spinnerMoneda.setSelection(0);
        binding.editCantidad.setText("0");
        binding.editMotivo.setText("");
        binding.checkIngreso.setChecked(false);
        binding.txtTagSeleccionados.setText(getResources().getString(R.string.sin_seleccionados));

        tags.clear();
        monedas.clear();
        cantidades.clear();

        Cursor data = dataBase.getMonedasByUserId(UsuarioSingleton.getInstance().getID());
        while (data.moveToNext()) {
            monedas.add(data.getString(1));
            cantidades.add(data.getFloat(2));
            simbolos.add(data.getString(3));
        }
        llenarLayoutMonedas();

        binding.scrollview.fullScroll(ScrollView.FOCUS_UP);

    }


    private void popupCerrarSesion() {
        AlertDialog.Builder builder = new AlertDialog.Builder(requireContext());
        builder.setTitle(getResources().getString(R.string.salir));
        builder.setMessage(getResources().getString(R.string.quieres_cerrar_sesion));
        DialogInterface.OnClickListener c = (dialogInterface, i) -> {
            Preferences.deleteFiltros(requireContext());
            Preferences.deleteUser(requireContext());
            Preferences.deletePreferencesGastoPendiente(requireContext());
            Preferences.deletePreferencesEdicionGasto(requireContext());
            Intent intent = new Intent(requireContext(), ActivityLogin.class);
            startActivity(intent);
            requireActivity().finish();
        };
        builder.setPositiveButton(getResources().getString(R.string.si), c);
        builder.setNegativeButton(getResources().getString(R.string.no), null);
        builder.setCancelable(true);

        AlertDialog dialog = builder.create();
        dialog.show();
    }


}

File: app/src/main/java/mis/finanzas/diarias/fragments/PrincipalFragment.kt
package mis.finanzas.diarias.fragments

import androidx.annotation.RequiresApi
import android.os.Build
import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import com.example.taskdone.R
import android.view.Gravity
import android.text.TextWatcher
import android.text.Editable
import mis.finanzas.diarias.DatePickerFragment
import android.view.View
import android.widget.*
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import com.example.taskdone.databinding.FragmentFinanzasPrincipalBinding
import mis.finanzas.diarias.Utils
import mis.finanzas.diarias.activities.ActivityFinanzas
import mis.finanzas.diarias.model.Currency
import mis.finanzas.diarias.model.Record
import mis.finanzas.diarias.viewmodels.DatabaseViewModel
import mis.finanzas.diarias.viewmodels.DatabaseViewmodelFactory
import mis.finanzas.diarias.viewmodels.RecordViewModel
import java.text.ParseException
import java.util.*

@RequiresApi(api = Build.VERSION_CODES.O)
class PrincipalFragment : Fragment() {
    private lateinit var binding: FragmentFinanzasPrincipalBinding
    private val databaseViewModel: DatabaseViewModel by activityViewModels{DatabaseViewmodelFactory(requireContext())}
    private val recordViewModel: RecordViewModel by activityViewModels()

    var tags: ArrayList<String>? = null

    @SuppressLint("SetTextI18n", "UseCompatLoadingForDrawables", "RestrictedApi")
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentFinanzasPrincipalBinding.inflate(inflater, container, false)
        databaseViewModel.getAllCurrency()

        setObservers()

        binding.spinnerMoneda.background =
            resources.getDrawable(R.drawable.fondo_blanco_redondeado)
        binding.spinnerMoneda.gravity = Gravity.CENTER

        binding.agregarTag.setOnClickListener {
            run {
            (activity as ActivityFinanzas).updateFragment(R.id.tagsFragment)
            findNavController().navigate(R.id.tagsFragment)
            }
        }

        binding.ok.setOnClickListener {
                addRecord()
        }
        tags = ArrayList()
        if (arguments != null && requireArguments().getString("scroll_tags") != null) {
            scrollToTag()
        }
        binding.editFecha.setOnClickListener { showDatePicker() }
        binding.agregarMoneda.setOnClickListener {
            (activity as ActivityFinanzas).updateFragment(R.id.crearMonedaFragment)
            findNavController().navigate(R.id.crearMonedaFragment)
        }
        binding.editCantidad.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                verificarCantidad(charSequence.toString())
                recordViewModel.setAmount(
                    charSequence.toString().let {
                        if (it != "") it.toInt() else 0
                    })
            }
            override fun afterTextChanged(editable: Editable) {//
            }
        })
        binding.editMotivo.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                recordViewModel.setReason(binding.editMotivo.text.toString())
            }

            override fun afterTextChanged(editable: Editable) {//
            }
        })
        //cargarGastoPendiente()
        binding.ayuda.setOnClickListener {
            Utils.popupAyuda(
                requireContext(), requireActivity(), ArrayList(
                    Arrays.asList(
                        resources.getString(R.string.ayuda_principal_1),
                        resources.getString(R.string.ayuda_principal_2)
                    )
                )
            )
        }
        reloadData()
        return binding.root
    }

    private fun setObservers(){
        databaseViewModel.currencies.observe(viewLifecycleOwner) { currencyList ->
            if (currencyList.map { it.name }.isEmpty()) {

                (activity as ActivityFinanzas).updateFragment(R.id.crearMonedaFragment)
                findNavController().navigate(R.id.crearMonedaFragment)
            }
            if(binding.spinnerMoneda.adapter == null || binding.spinnerMoneda.adapter?.count != currencyList.count()) {
                val spinnerArrayAdapter: ArrayAdapter<String?> = object : ArrayAdapter<String?>(
                    requireActivity(),
                    android.R.layout.simple_spinner_dropdown_item,
                    currencyList.map { it.name } as List<String?>
                ) {
                    @RequiresApi(api = Build.VERSION_CODES.O)
                    override fun getView(
                        position: Int,
                        convertView: View?,
                        parent: ViewGroup
                    ): View {
                        val v = super.getView(position, convertView, parent)
                        (v as TextView).textSize = 20f
                        v.textAlignment = View.TEXT_ALIGNMENT_CENTER
                        return v
                    }
                }
                spinnerArrayAdapter.setDropDownViewResource(R.layout.spinner_dropdown)
                binding.spinnerMoneda.adapter = spinnerArrayAdapter
                binding.spinnerMoneda.setSelection(currencyList.indexOf(recordViewModel.getCurrency()))

                binding.spinnerMoneda.onItemSelectedListener =
                    object : AdapterView.OnItemSelectedListener {
                        override fun onItemSelected(
                            adapterView: AdapterView<*>?,
                            view: View?,
                            i: Int,
                            l: Long
                        ) {
                            recordViewModel.setCurrency(currencyList[i])
                        }

                        override fun onNothingSelected(adapterView: AdapterView<*>?) {//
                        }
                    }
            }

            while (binding.currencyLayout.childCount != 0) {
                binding.currencyLayout.removeViewAt(0)
            }
            val inflater = requireActivity().layoutInflater
            for (x in currencyList.map { it.name }.indices) {
                @SuppressLint("InflateParams") val view =
                    inflater.inflate(R.layout.item_moneda_cantidad, null)
                @SuppressLint("CutPasteId") val moneda = view.findViewById<TextView>(R.id.name)
                @SuppressLint("CutPasteId") val cantidad = view.findViewById<TextView>(R.id.cantidad)
                moneda.text = currencyList.map { it.name }[x]
                cantidad.text =
                    currencyList.map { it.symbol }[x] + " " + Utils.formatoCantidad(currencyList.map { it.amount }[x])
                binding.currencyLayout.addView(view)
            }
        }
        binding.checkIngreso.setOnCheckedChangeListener { _: CompoundButton?, _: Boolean ->
            recordViewModel.setIncome(!recordViewModel.getIncome())
        }
    }

    private fun reloadData(){
        databaseViewModel.refreshAllCurrency()
        recordViewModel.let {
            binding.editFecha.text = it.getDate()
            binding.editCantidad.setText(it.getAmount().toString())
            binding.editMotivo.setText( it.getReason())
            binding.txtTagSeleccionados.text = it.getTagsString(resources)
            binding.checkIngreso.isChecked = it.getIncome()
        }
    }
    @SuppressLint("SetTextI18n")
    private fun cleanAndUpdate() {
        //binding.spinnerMoneda.setSelection(0)
        recordViewModel.setAmount(0)
        recordViewModel.setReason("")
        recordViewModel.setIncome(false)
        recordViewModel.setSelectedTags(arrayListOf())
        reloadData()
        binding.scrollview.fullScroll(ScrollView.FOCUS_UP)
    }


    private fun scrollToTag() {
        binding.scrollview.post { binding.scrollview.fullScroll(ScrollView.FOCUS_DOWN) }
    }

    private fun verificarCantidad(charSequence: String) {
        if (charSequence.length > 1 && charSequence.startsWith("0")) {
            binding.editCantidad.setText(charSequence.substring(1))
            binding.editCantidad.setSelection(binding.editCantidad.length())
        }
    }

    fun addRecord() {
        // FECHA DEL FUTURO NO SE PERMITE
        if (Utils.fechaMayorQueHoy(binding.editFecha.text.toString())) {
            Toast.makeText(requireContext(), R.string.error_fecha_futura, Toast.LENGTH_SHORT).show()
            return
        }
        if(recordViewModel.getAmount() == 0) {
            Toast.makeText(requireContext(), R.string.error_cantidad_null, Toast.LENGTH_SHORT).show()
            return
        }
        val record = Record(recordViewModel.getDate(),
                            recordViewModel.getAmount(),
                            recordViewModel.getReason(),
                            recordViewModel.getIncome(),
                            recordViewModel.getCurrency().id)
        databaseViewModel.addRecord(record)
        Toast.makeText(requireContext(), R.string.guardado_exito, Toast.LENGTH_SHORT)
                .show()

        cleanAndUpdate()
    }

    private fun showDatePicker() {
        DatePickerFragment.newInstance { _: DatePicker?, year: Int, month: Int, day: Int ->
            val selectedDate =
                year.toString() + "/" + Utils.twoDigits(month + 1) + "/" + Utils.twoDigits(day)
            val h = Calendar.getInstance()
            h[year, month] = day
            h.add(Calendar.DAY_OF_YEAR, 1)
            binding.editFecha.text = selectedDate
            recordViewModel.setDate(selectedDate)
        }.show(
            Objects.requireNonNull(requireActivity()).supportFragmentManager,
            "datePicker"
        )
    }


}
